docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,214 @@
+
+The new Go SDK also separates each Azure service (containerservice, network, etc.) into distinct Go modules, where each module version maps to a single Azure API version. Modules for stable APIs use semantic versions like "v1.0.0" while for preview APIs they have pre-release versions like "v1.0.0-beta.1" with the same major version as the stable API. This would make it difficult to use different API versions side-by-side, like a preview and stable version for https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2625, since the Go toolchain likely will never natively support importing two different minor/patch versions of the same module: https://github.com/golang/go/issues/31578. ASO instead exposes multiple Azure API versions in each of its releases with different Kubernetes API versions of its types, so CAPZ can use a single version of ASO to gain access to several Azure API versions more easily than with the SDK directly.
+
+### Crossplane
+
+Another alternative is [Crossplane](https://www.crossplane.io/), a CNCF project which is a set of Kubernetes controllers and Custom Resource Definitions like ASO that define and operate on resources for Azure and other cloud providers. A number of factors indicate that ASO may be a better fit for CAPZ:
+- CRD definitions generated from the Azure Terraform provider maintained by
+  HashiCorp, which may not keep up with changes as fast as the Azure API swagger
+  specification from which ASO generates its definitions
+- Still using the [deprecated Azure Go SDK](https://github.com/crossplane/crossplane/blob/08f11debb66cdf10ff0ddd14f1ee23aa9b4489b1/go.mod#L46-L55), no apparent plans to update
+- [Open issue](https://github.com/upbound/provider-azure/issues/94) to support workload identity, which ASO already handles
+
+## Upgrade Strategy
+
+There are two main upgrade paths CAPZ needs to consider: the first between a version of CAPZ that does not use ASO to one that does, and the second between two versions of CAPZ that use different versions of ASO.
+
+During the initial adoption of ASO, CAPZ's ASO-based [reconciliation flow](#reconciliation) will automatically adopt resources previously managed by CAPZ based on the `sigs.k8s.io_cluster-api-provider-azure_cluster_*` tag on the Azure resource.
```

Comment:
One potential hiccup with this plan is that managed cluster resources in Azure do not receive this tag at the moment. I've been working on a fix for #3009 that includes adding the tag, so this would work assuming the "old" CAPZ version includes that fix, but not versions older than that. Since BYO AKS has never been implemented, one option might be to add a managed cluster-specific hack here to always adopt those resources.

Reply:
Are there any other resources that don't have this tag? Some resource types such as subnets don't support tags at all AFAIK, how would we handle those?

> Since BYO AKS has never been implemented, one option might be to add a managed cluster-specific hack here to always adopt those resources.

agreed

Reply:
> Are there any other resources that don't have this tag?

Based on whether or not `infrav1.Build` is used, it looks like these services do not add the tag:
- agentpools
- disks
- inboundnatrules
- managedclusters
- roleassignments
- subnets
- vmextensions
- vnetpeerings

Of those, it looks possible to add the tag for:
- agentpools
- disks
- managedclusters
- vmextensions

> Some resource types such as subnets don't support tags at all AFAIK, how would we handle those?

I imagine we'd want the same behavior as each service's `IsManaged()` method, where subnets are managed if the owning vnet is managed and the other three that don't support tags (inboundnatrules, roleassignments, vnetpeerings) are always assumed to be managed. So a similar shortcut can maybe be applied to those.